modules/gapi/include/opencv2/gapi/infer.hpp

```diff
@@ -395,35 +445,178 @@ struct Generic { };
             });
 
-    call->setArgs(std::move(input_args));
-    call->params() = InOutInfo{input_names, {}};
+    call->setArgs(std::move(args));
+    call->params() = InOutInfo{std::move(names), {}};
+
+    return call;
+}
+
+inline void unpackBlobs(const GInferInputs::Map& blobs,
+                        std::vector<cv::GArg>& args,
+                        std::vector<std::string>& names,
+                        cv::GKinds& kinds)
+{
+    for (auto&& p : blobs) {
+        names.emplace_back(p.first);
+        switch (p.second.index()) {
+            case cv::GInferInputs::StorageT::index_of<cv::GMat>():
+                args.emplace_back(cv::util::get<cv::GMat>(p.second));
+                kinds.emplace_back(cv::detail::OpaqueKind::CV_MAT);
+                break;
+            case cv::GInferInputs::StorageT::index_of<cv::GFrame>():
+                args.emplace_back(cv::util::get<cv::GFrame>(p.second));
+                kinds.emplace_back(cv::detail::OpaqueKind::CV_UNKNOWN);
+                break;
+            default:
+                GAPI_Assert(false);
+        }
+    }
+}
+
+inline void unpackBlobs(const GInferInputs::Map& blobs,
+                        std::vector<cv::GArg>& args,
+                        std::vector<std::string>& names,
+                        cv::GKinds& kinds)
+{
+    for (auto&& p : blobs) {
+        names.emplace_back(p.first);
+        switch (p.second.index()) {
+            case cv::GInferInputs::StorageT::index_of<cv::GMat>():
+                args.emplace_back(cv::util::get<cv::GMat>(p.second));
+                kinds.emplace_back(cv::detail::OpaqueKind::CV_MAT);
+                break;
+            case cv::GInferInputs::StorageT::index_of<cv::GFrame>():
+                args.emplace_back(cv::util::get<cv::GFrame>(p.second));
+                kinds.emplace_back(cv::detail::OpaqueKind::CV_UNKNOWN);
+                break;
+            default:
+                GAPI_Assert(false);
+        }
+    }
+}
+
+template<typename T = Generic> GInferOutputs
+infer(const std::string& tag, const GInferInputs& inputs)
+{
+    std::vector<cv::GArg> args;
+    std::vector<std::string> names;
+    cv::GKinds kinds;
+
+    unpackBlobs(inputs.getBlobs(), args, names, kinds);
+
+    auto call = makeCall<GInferBase>(tag,
+                                     std::move(args),
+                                     std::move(names),
+                                     std::move(kinds));
 
     return GInferOutputs{std::move(call)};
 }
 
+template <typename InferType>
+struct InferROITraits;
+
+template <>
+struct InferROITraits<GInferROIBase>
+{
+    using outType = cv::GInferOutputs;
+    using inType  = cv::GOpaque<cv::Rect>;
+};
+
+template <>
+struct InferROITraits<GInferListBase>
+{
+    using outType = cv::GInferListOutputs;
+    using inType  = cv::GArray<cv::Rect>;
+};
+
+template<typename InferType> 
+typename InferROITraits<InferType>::outType
+inferROI(const std::string& tag,
```

Comment:
Does it look ugly ? Maybe move some stuff into `detail` or `infer_priv.hpp` ?

Reply:
Absolutely agree. All implementation details which are not a part of the end API should be placed under detail namespace at least. Maybe it's also worth to get rid of templates in the API part by creating the according overloads with supported types, but maybe it's not. When the details are under detail namespace there will be more illustrative view of that can and should be hidden and that's not

Reply:
Done